### PR TITLE
Implement measurements using decorator pattern

### PR DESCRIPTION
**Pull request by bassosimone**
This also includes a new experiment called urlgetter that can
be used to perform URL measurements like:

```
./miniooni -i http://www.kernel.org urlgetter

./miniooni -OTLSServerName=x.org -i http://www.kernel.org urlgetter

./miniooni -OResolverURL=https://cloudflare-dns.com/dns-query \
           -i http://www.kernel.org urlgetter
```

Part of https://github.com/ooni/probe-engine/issues/509

### CODE REVIEW

File: domain-check.bash [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#
+# Inspired by great chats and great code. Main source of ispiration is
+# currently https://github.com/Jigsaw-Code/net-analysis/blob/master/netanalysis/blocktest/measure.sh
+
+function fatal() {
+  echo "FATAL: $@" 1>&2
+  exit 1
+}
+
+if [ $# -ne 1 ]; then
```
Comment (fortuna) [inline domain-check.bash:16]: Prefer `[[]]` to `[]`: https://google.github.io/styleguide/shellguide.html#s6.3-tests

BTW, that's a pretty good style guide. You may want to use it if you don't already have one.

File: domain-check.bash [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+fi
+domain=$1
+
+function require() {
+  if ! [ -x "$(command -v $1)" ]; then
+    fatal "$1 not installed, please run: $2"
+  fi
+}
+
+require jq "sudo apt install jq"
```
Comment (fortuna) [inline domain-check.bash:27]: I recommend you put all statements that are executed inside a `main()` function, so it's easier to read and understand. Then you call `main "$@"` at the bottom.

File: domain-check.bash [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+function getipv4list() {
+  tail -n1 report.jsonl|jq -r ".test_keys.queries|.[]|select(.hostname==\"$1\")|select(.query_type==\"A\")|.answers|.[].ipv4"|sort
+}
+
+function getbody() {
+  # Implementation note: requests stored in LIFO order
+  tail -n1 report.jsonl|jq -r ".test_keys.requests|.[0]|.response.body"
+}
+
+testdomain=${MINIOONI_TEST_DOMAIN:-example.org}
```
Comment (fortuna) [inline domain-check.bash:67]: You may want to organize the code below in functions, so it's easier to read.

File: domain-check.bash [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+if [ "$(getfailure)" = "null" ]; then
+  output "MINIOONI_DNS_INJECTION=1"
+fi
+
+log "* resolving $domain using the system resolver"
+# Implementation note: this saves the IPs _and_ records the error.
+# XXX: maybe make these two checks separate checks for now?
+urlgetter -OResolverURL=system:/// -ORejectDNSBogons=true -i dnslookup://$domain
+if [ "$(getfailure)" = "dns_bogon_error" ]; then
+  output "MINIOONI_DNS_BOGONS=1"
```
Comment (fortuna) [inline domain-check.bash:104]: This doesn't really explain the censorship mechanism.
Was the bogon injected or did the resolver lie?
That's what matters the most. The bogon detection is just a way to validate the IP address, but the result should be DNS lie or DNS injection.

File: domain-check.bash (hunk 1 above)
Comment (bassosimone) [reply domain-check.bash:16]: Thanks for letting me know about this style guide, I was not aware of that!

File: domain-check.bash (hunk 2 above)
Comment (bassosimone) [reply domain-check.bash:27]: ACK

File: domain-check.bash (hunk 3 above)
Comment (bassosimone) [reply domain-check.bash:67]: Good point

File: domain-check.bash (hunk 4 above)
Comment (bassosimone) [reply domain-check.bash:104]: I agree. I think this script was an initial experiment to understand how an higher level logic for running measurements (i.e. an improved web connectivity) could work. For this reason, I didn't bother with implementing an algorithm for providing output to users of the script. The goal is to understand a good algorithm and then implement it in Go.

**Comment by bassosimone** [comment]
This pull request has served its purpose of providing us with a PoC regarding how to implement better measurements. We should spend some time as part of Sprint 21 trying to understand if there is still valuable information that we can extract from here as a set of issues, or whether we can simply close this PR, delete the associated branch, and move on.

**Comment by bassosimone** [comment]
Effort is `1` because we need to take a look.

**Comment by bassosimone** [comment]
So, the remaining part of this diff that has not been merged into master only exposes my lack of good shell writing skills and does not otherwise carry much meaningful content. The problem of improving Web Connectivity has not been solved yet but the place where to do that is clearly not this pull request. This is for many reasons, including that I'd like the improved domain checking code to be implemented in Go and to be callable as a OONI experiment. So, safe to close this PR.

**Pull request closed without merging**